applications/StructuralMechanicsApplication/custom_constitutive/constitutive_laws_integrators/d+d-constitutive_law_integrators/generic_compression_constitutive_law_integrator.h

```diff
@@ -0,0 +1,331 @@
+    {
+        return *this;
+    }
+
+    /// Destructor
+    virtual ~GenericCompressionConstitutiveLawIntegratorDplusDminusDamage()
+    {
+    }
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief This method integrates the predictive stress vector with the CL using linear or exponential softening
+     * @param PredictiveStressVector The predictive stress vector
+     * @param UniaxialStress The equivalent uniaxial stress
+     * @param Damage The internal variable of the damage model
+     * @param Threshold The maximum uniaxial stress achieved previously
+     * @param rValues Parameters of the constitutive law
+     * @param CharacteristicLength The equivalent length of the FE
+     */
+    static void IntegrateStressVector(
+        array_1d<double, VoigtSize>& rPredictiveStressVector,
+        const double UniaxialStress,
+        double& rDamage,
+        double& rThreshold,
+        ConstitutiveLaw::Parameters& rValues,
+        const double CharacteristicLength
+        )
+    {
+        const Properties& r_material_properties = rValues.GetMaterialProperties();
+        const int softening_type = r_material_properties[SOFTENING_TYPE];
+        double damage_parameter;
+        TYieldSurfaceType::CalculateDamageParameter(rValues, damage_parameter, CharacteristicLength);
+
+        switch (softening_type)
+        {
+        case static_cast<int>(SofteningType::Linear):
+            CalculateLinearDamage(UniaxialStress, rThreshold, damage_parameter, CharacteristicLength, rValues, rDamage);
+            break;
+        case static_cast<int>(SofteningType::Exponential):
+            CalculateExponentialDamage(UniaxialStress, rThreshold, damage_parameter, CharacteristicLength, rValues, rDamage);
+            break;
+        default:
+            KRATOS_ERROR << "SOFTENING_TYPE not defined or wrong..." << softening_type << std::endl;
+            break;
+        }
+        rPredictiveStressVector *= (1.0 - rDamage);
+    }
+
+    /**
+     * @brief This method returns the initial uniaxial stress threshold
+     * @param rThreshold The uniaxial stress threshold
+     * @param rValues Parameters of the constitutive law
+     */
+    static void GetInitialUniaxialThreshold(
+        ConstitutiveLaw::Parameters& rValues,
+        double& rThreshold
+        )
+    {
+        // This is done to allow tension-driven yields to work as compression yields
+        if (YieldSurfaceType::IsWorkingWithTensionThreshold()) {
+            ConstitutiveLaw::Parameters modified_ones = rValues;
+            const double yield_compression = modified_ones.GetMaterialProperties()[YIELD_STRESS_COMPRESSION];
+            Properties material_props = modified_ones.GetMaterialProperties();
+            material_props.SetValue(YIELD_STRESS_TENSION, yield_compression);

```

Comment:
It may be confussing but, in my opinion, necessary... The problem arises when you want use a Tensile yield surface like VonMises as a compression yield surface (if you see the von_mises_yield_surface.h it always compare with the YIELD_STRESS_TENSION).

Reply:
But will it vice versa affect the yield stress tension for the real tension case? In sence of load - unloading - but also the directions. Or is this local for each Yield Surface?

Reply:
Nope! llok that only corrects the case of a tensile yield surf working inside the compressive integrator... If the same VonMises is working as a tensile yield surf, it is templated inside the TensionIntegrator where there is no modification of the yield stress.

Reply:
ah alright. Do can we do the same for the Fracture Energy (That we set the FRACTURE_ENERGY to FRACTURE_ENERGY_COMPRESSION) -->see line 133.

Reply:
yes!! tomorrow I'll think about that and I'll comment the changes!